win32ss/user/ntuser/ghost.c

```diff
@@ -4,42 +4,109 @@
  * PURPOSE:     Window ghosting feature
  * COPYRIGHT:   Copyright 2018 Katayama Hirofumi MZ (katayama.hirofumi.mz@gmail.com)
  */
-
 #include <win32k.h>
 #include "ghostwnd.h"
 
-#define NDEBUG
-#include <debug.h>
-
 DBG_DEFAULT_CHANNEL(UserInput);
 
 static UNICODE_STRING GhostClass = RTL_CONSTANT_STRING(GHOSTCLASSNAME);
-static UNICODE_STRING GhostProp = RTL_CONSTANT_STRING(GHOST_PROP);
+static UNICODE_STRING TrayWndClass = RTL_CONSTANT_STRING(L"Shell_TrayWnd");
+static UNICODE_STRING ProgmanWndClass = RTL_CONSTANT_STRING(L"Progman");
 
 PTHREADINFO gptiGhostThread = NULL;
+PKEVENT gpGhostStartupEvent = NULL;
 
-/*
- * GhostThreadMain
- *
- * Creates ghost windows and exits when no non-responsive window remains.
- */
-VOID NTAPI
-UserGhostThreadEntry(VOID)
+// linked list node
+typedef struct GHOST_ENTRY
 {
-    TRACE("Ghost thread started\n");
+    LIST_ENTRY ListEntry;
+    HWND hwndTarget;
+    HWND hwndGhost;
+} GHOST_ENTRY, *PGHOST_ENTRY;
 
-    UserEnterExclusive();
+LIST_ENTRY gGhostListHead = { &gGhostListHead, &gGhostListHead };
 
-    gptiGhostThread = PsGetCurrentThreadWin32Thread();
+PGHOST_ENTRY FASTCALL
+IntFindGhostEntryFromTargetWindow(HWND hwndTarget)
+{
+    PLIST_ENTRY CurrentEntry, ListHead = &gGhostListHead;
+    PGHOST_ENTRY pEntry;
 
-    //TODO: Implement. This thread should handle all ghost windows and exit when no ghost window is needed.
+    for (CurrentEntry = ListHead->Flink;
+         CurrentEntry != ListHead;
+         CurrentEntry = CurrentEntry->Flink)
+    {
+        pEntry = CONTAINING_RECORD(CurrentEntry, GHOST_ENTRY, ListEntry);
 
-    gptiGhostThread = NULL;
+        if (pEntry->hwndTarget == hwndTarget)
+            return pEntry;
+    }
 
-    UserLeave();
+    return NULL;
+}
+
+PGHOST_ENTRY FASTCALL
+IntFindGhostEntryFromGhostWindow(HWND hwndGhost)
+{
+    PLIST_ENTRY CurrentEntry, ListHead = &gGhostListHead;
+    PGHOST_ENTRY pEntry;
+
+    for (CurrentEntry = ListHead->Flink;
+         CurrentEntry != ListHead;
+         CurrentEntry = CurrentEntry->Flink)
+    {
+        pEntry = CONTAINING_RECORD(CurrentEntry, GHOST_ENTRY, ListEntry);
+
+        if (pEntry->hwndGhost == hwndGhost)
+            return pEntry;
+    }
+
+    return NULL;
+}
+
+__inline PGHOST_ENTRY FASTCALL
+IntCreateGhostEntry(HWND hwndTarget, HWND hwndGhost)
+{
+    PGHOST_ENTRY pEntry;
+    pEntry = ExAllocatePoolWithTag(NonPagedPool, sizeof(*pEntry), USERTAG_GHOST);
+    if (!pEntry)
+    {
+        ERR("ExAllocatePoolWithTag failed");
+        return NULL;
+    }
+
+    pEntry->hwndTarget = hwndTarget;
+    pEntry->hwndGhost = hwndGhost;
+    return pEntry;
+}
+
+__inline BOOL FASTCALL
+IntAddGhostEntry(HWND hwndTarget, HWND hwndGhost)
```

Comment:
It's a style choice, but to me IntCreateGhostEntry/IntAddGhostEntry seems small enough to be one function.